src/Gravity.hpp

```diff
@@ -0,0 +1,492 @@
+	/// @param grav_vector  MultiFab containing gravity vector
+	/// @param time         Current time
+	///
+	void get_new_grav_vector(int level, amrex::MultiFab &grav_vector, amrex::Real time);
+
+	///
+	/// Test whether using the edge-based gradients
+	/// to compute Div(Grad(Phi)) satisfies Lap(phi) = RHS
+	///
+	/// @param bx          box
+	/// @param rhs         right-hand-side
+	/// @param ecx         gradients wrt x
+	/// @param ecy         gradients wrt y
+	/// @param ecz         gradients wrt z
+
+	static void test_residual(const amrex::Box &bx, amrex::Array4<amrex::Real> const &rhs, amrex::Array4<amrex::Real> const &ecx,
+				  amrex::Array4<amrex::Real> const &ecy, amrex::Array4<amrex::Real> const &ecz,
+				  amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> dx);
+
+	///
+	/// @param level
+	///
+	void test_level_grad_phi_prev(int level);
+
+	///
+	/// @param level
+	///
+	void test_level_grad_phi_curr(int level);
+
+	///
+	/// @param level
+	///
+	void test_composite_phi(int level);
+
+	///
+	/// @param level    Level index
+	/// @param is_new   Use new state data (1) or old state data (0)?
+	///
+	void average_fine_ec_onto_crse_ec(int level, int is_new);
+
+	///
+	/// Implement multipole boundary conditions
+	///
+	/// @param crse_level
+	/// @param fine_level
+	/// @param Rhs
+	/// @param phi
+	///
+	void fill_multipole_BCs(int crse_level, int fine_level, const amrex::Vector<amrex::MultiFab *> &Rhs, amrex::MultiFab &phi);
+
+	///
+	/// Initialize multipole gravity
+	///
+	void init_multipole_grav();
+
+	///
+	/// Make multigrid boundary conditions
+	///
+	void make_mg_bc();
+
+	///
+	/// Pointers to amr,amrlevel.
+	///
+	//  amrex::Amr *parent;
+	//  amrex::Vector<amrex::AmrLevel *> LevelData;
+
+	///
+	/// Pointer to AMRSimulation<problem_t>
+	///
+	AMRSimulation<T> *sim;
+
+	///
+	/// MultiFabs for potential, acceleration
+	///
+	amrex::Vector<amrex::MultiFab> phi_old_;
+	amrex::Vector<amrex::MultiFab> phi_new_;
+	amrex::Vector<amrex::MultiFab> g_old_;
+	amrex::Vector<amrex::MultiFab> g_new_;
+
+	///
+	/// Pointers to grad_phi at previous and current time
+	///
+	amrex::Vector<amrex::Vector<std::unique_ptr<amrex::MultiFab>>> grad_phi_curr;
+	amrex::Vector<amrex::Vector<std::unique_ptr<amrex::MultiFab>>> grad_phi_prev;
+
+	///
+	/// MultiFabs for composite-level corrections
+	///
+	amrex::Vector<std::unique_ptr<amrex::MultiFab>> corr_phi_;
+	amrex::Vector<amrex::Vector<std::unique_ptr<amrex::MultiFab>>> corr_grad_phi_;
+
+	///
+	/// BoxArray at each level
+	///
+
+	///
+	/// Absolute tolerance on each level
+	///
+	amrex::Vector<amrex::Real> abs_tol;
+
+	///
+	/// Relative tolerance on each level
+	///
+	amrex::Vector<amrex::Real> rel_tol;
+
+	///
+	/// Resnorm at each level
+	///
+	amrex::Vector<amrex::Real> level_solver_resnorm;
+
+	///
+	/// Maximum value of the RHS (used for obtaining absolute tolerances)
+	///
+	amrex::Real max_rhs;
+
+	const amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> coordCenter;
```

Comment:
We don't need any of these variables anymore, since OpenBC computes the boundary conditions self-consistently. So all of the multipole stuff can be deleted.

Reply:
If I delete:

```
namespace multipole
{
const int lnum_max = 30;

constexpr amrex::Real volumeFactor = 1.0;
constexpr amrex::Real parityFactor = 1.0;

amrex::Array1D<bool, 0, 2> constexpr doSymmetricAddLo = {false};
amrex::Array1D<bool, 0, 2> constexpr doSymmetricAddHi = {false};
bool constexpr doSymmetricAdd = false;

amrex::Array1D<bool, 0, 2> constexpr doReflectionLo = {false};
amrex::Array1D<bool, 0, 2> constexpr doReflectionHi = {false};

extern AMREX_GPU_MANAGED amrex::Real rmax;

extern AMREX_GPU_MANAGED amrex::Array2D<amrex::Real, 0, lnum_max, 0, lnum_max> factArray;
extern AMREX_GPU_MANAGED amrex::Array1D<amrex::Real, 0, lnum_max> parity_q0;
extern AMREX_GPU_MANAGED amrex::Array2D<amrex::Real, 0, lnum_max, 0, lnum_max> parity_qC_qS;
} // namespace multipole
```

I get lots of errors :
```
/Users/psharda/Documents/quokka/src/GravityBC_util.hpp:126:39: error: use of undeclared identifier 'multipole'
                                dQL0 = legPolyL * rho_r_L * vol * multipole::volumeFactor;
                                                                  ^
/Users/psharda/Documents/quokka/src/GravityBC_util.hpp:128:20: error: use of undeclared identifier 'multipole'
                                        dQL0 = dQL0 * multipole::parity_q0(l);
                                        
```

Reply:
So do I also delete the `multipole_add` function in `GravityBC_util.hpp`

Reply:
You can delete both `GravityBC.hpp` and `GravityBC_util.hpp`.